optimade/adapters/base.py

```diff
@@ -0,0 +1,115 @@
+import re
+from typing import Union, Dict, Callable, Any
+
+from optimade.models import EntryResource
+
+
+class EntryAdapter:
+    """Base class for lazy resource entry adapters
+    :param entry: JSON OPTIMADE single resource entry.
+    """
+
+    ENTRY_RESOURCE: EntryResource = EntryResource
+    _type_converters: Dict[str, Callable] = {}
+
+    def __init__(self, entry: dict):
+        self._entry = None
+        self._converted = {}
+
+        self.entry = entry
+
+        # Note that these return also the default values for otherwise non-provided properties.
+        self._common_converters = {
+            "json": self.entry.json,  # Return JSON serialized string, see https://pydantic-docs.helpmanual.io/usage/exporting_models/#modeljson
+            "dict": self.entry.dict,  # Return Python dict, see https://pydantic-docs.helpmanual.io/usage/exporting_models/#modeldict
+        }
+
+    @property
+    def entry(self):
+        """Get OPTIMADE entry"""
+        return self._entry
+
+    @entry.setter
+    def entry(self, value: dict):
+        """Set OPTIMADE entry
+        If already set, print that this can _only_ be set once.
+        """
+        if self._entry is None:
+            self._entry = self.ENTRY_RESOURCE(**value)
+        else:
+            print("entry can only be set once and is already set.")
+
+    def convert(self, format: str) -> Any:
+        """Convert OPTIMADE entry to desired format"""
+        if (
+            format not in self._type_converters
+            and format not in self._common_converters
+        ):
+            raise AttributeError(
+                f"Non-valid entry type to convert to: {format}\n"
+                f"Valid entry types: {tuple(self._type_converters.keys()) + tuple(self._common_converters.keys())}"
+            )
+
+        if self._converted.get(format, None) is None:
+            if format in self._type_converters:
+                self._converted[format] = self._type_converters[format](self.entry)
+            else:
+                self._converted[format] = self._common_converters[format]()
+
+        return self._converted[format]
+
+    def __getattr__(self, name: str) -> Any:
+        """Get converted entry or attribute from OPTIMADE entry
+        Order:
+        - Try to return converted entry if using `get_<_type_converters value>`.
+        - Try to return OPTIMADE ENTRY_RESOURCE attribute.
+        - Try to return OPTIMADE ENTRY_RESOURCE.attributes attribute.
+        - Raise AttributeError
+        """
+        # get_<entry_type>
```

Comment:
Minor quibble with this `adapter.get_format` nomenclature is that I would expect e.g. `get_json` to be a callable, do you think something like `adapter.as_json` is a better name?

Reply:
Yeah, that might be a good idea. I feel the same way as you about `get`, so `as` seems the better choice.